src/QuartoPlayerAgent.java

```diff
@@ -52,25 +44,95 @@ public static void main(String[] args) {
+        int maxSearchableNodes = NODES_PER_SECOND * (this.timeLimitForResponse / 1000);
+        while(totalNodes <= maxSearchableNodes) {
+            System.out.println("--nodes in generation: " + QuartoPlayerAgent.curState.calcNodesInGeneration(depthLimit));
+            depthLimit += 1;
+            totalNodes += QuartoPlayerAgent.curState.calcNodesInGeneration(depthLimit);
+        }
+
+        if(depthLimit > QuartoPlayerAgent.maxDepth) {
+            QuartoPlayerAgent.maxDepth = depthLimit;
+            System.out.printf("--MaxDepth so far: %d\n", QuartoPlayerAgent.maxDepth);
+        }
+
+        return depthLimit - 1;
     }
+
     /**
-     * @param        curState
-     * @param        levelsLeft
+     * @param curState
+     * @param levelsLeft
+     * @param limboPiece
+     * @param rootDepth The current root nodes depth in relation to the original tree. Used to ensure unique evaluations.
      */
-    private void searchGameTree(QuartoGameState curState, int levelsLeft)
+    private static void searchGameTree(QuartoGameState curState, QuartoPiece limboPiece, int levelsLeft, int rootDepth)
     {
         if(levelsLeft == 0 || curState.hasQuarto()) {
-            curState.evaluate();
+            //Reached max depth or a terminal winning node
+            curState.evaluate(rootDepth);
         }
         else {
-            for(QuartoGameState state : curState) {
-                if (state.value == null) {
+            QuartoGameTransitionGenerator gen = new QuartoGameTransitionGenerator(curState, limboPiece);
+            for(QuartoGameTransition transition : gen) {
+                QuartoGameState state = transition.toState;
+
+                System.out.println(state + " : " + state.getHash() + " : " + state.lastLevelExaminedFrom);
+                if (state.lastLevelExaminedFrom != rootDepth) {
+                    state.lastLevelExaminedFrom = rootDepth;
+
                     state.resetMinimax();
-                    this.searchGameTree(state, levelsLeft - 1);
-                    // adjust curState minimax values
+                    searchGameTree(state, transition.nextPiece, levelsLeft - 1, rootDepth);
+
+                    if (curState.isMaxState) {
+                        //Choose largest value
+                        if (state.value > Math.abs(curState.value)) {
+                            curState.value = curState.alpha = (state.value == 0) ? 0 : (Math.abs(state.value) - 1) * (Math.abs(state.value) / state.value);
+                            curState.bestTransition = transition;
+                        } else if (state.value == Math.abs(curState.value) && QuartoPlayerAgent.rand.nextInt(2)==0) {
+                            //50-50 chance we'll take a new equivalent transition
+                            curState.bestTransition = transition;
+                        }
+
+                        if (state.value + 1 < curState.beta) {
```

Comment:
It wouldn't be affected at all since we aren't using alpha or beta, we're doing the same thing for beta btw.

Either way this doesn't matter so I'll just merge the PR.